src/System.Security.Cryptography.X509Certificates/src/Internal/Cryptography/Pal.Unix/StorePal.cs

```
@@ -25,12 +30,123 @@ public static IStorePal FromCertificate(ICertificatePal cert)
+            }
+
+            if (StringComparer.Ordinal.Equals("Root", storeName))
+            {
+                return CloneStore(s_machineRootStore);
+            }
+
+            if (StringComparer.Ordinal.Equals("CA", storeName))
+            {
+                return CloneStore(s_machineIntermediateStore);
+            }
+
             throw new NotImplementedException();
         }
+
+        private static IStorePal CloneStore(X509Certificate2Collection seed)
+        {
+            X509Certificate2Collection coll = new X509Certificate2Collection();
+
+            foreach (X509Certificate2 cert in seed)
+            {
+                // Duplicate the certificate context into a new handle.
+                coll.Add(new X509Certificate2(cert.Handle));
+            }
+
+            return new OpenSslX509StoreProvider(coll);
+        }
+
+        private static void LoadMachineStores()
+        {
+            X509Certificate2Collection rootStore = new X509Certificate2Collection();
+
+            DirectoryInfo directoryInfo;
+
+            try
+            {
+                directoryInfo = new DirectoryInfo(Interop.NativeCrypto.GetX509RootStorePath());
+            }
+            catch (ArgumentException)
+            {
+                // If SSL_CERT_DIR is set to the empty string, or anything else which gives
+                // "The path is not of a legal form", then just call it a day.
+                s_machineRootStore = rootStore;
+                return;
+            }
+
+            if (!directoryInfo.Exists)
+            {
+                s_machineRootStore = rootStore;
+                return;
+            }
+
+            foreach (FileInfo file in directoryInfo.EnumerateFiles())
+            {
+                byte[] bytes;
+
+                try
+                {
+                    bytes = File.ReadAllBytes(file.FullName);
+                }
+                catch (IOException)
+                {
+                    continue;
```

I assume we're eating these exceptions to handle race conditions where the file gets deleted concurrently with this loop?